enum.c

```diff
@@ -871,49 +875,141 @@ ary_inject_op(VALUE ary, VALUE init, VALUE op)
 
 /*
  *  call-seq:
- *     enum.inject(initial, sym) -> obj
- *     enum.inject(sym)          -> obj
- *     enum.inject(initial) { |memo, obj| block }  -> obj
- *     enum.inject          { |memo, obj| block }  -> obj
- *     enum.reduce(initial, sym) -> obj
- *     enum.reduce(sym)          -> obj
- *     enum.reduce(initial) { |memo, obj| block }  -> obj
- *     enum.reduce          { |memo, obj| block }  -> obj
- *
- *  Combines all elements of <i>enum</i> by applying a binary
- *  operation, specified by a block or a symbol that names a
- *  method or operator.
- *
- *  The <i>inject</i> and <i>reduce</i> methods are aliases. There
- *  is no performance benefit to either.
- *
- *  If you specify a block, then for each element in <i>enum</i>
- *  the block is passed an accumulator value (<i>memo</i>) and the element.
- *  If you specify a symbol instead, then each element in the collection
- *  will be passed to the named method of <i>memo</i>.
- *  In either case, the result becomes the new value for <i>memo</i>.
- *  At the end of the iteration, the final value of <i>memo</i> is the
- *  return value for the method.
- *
- *  If you do not explicitly specify an <i>initial</i> value for <i>memo</i>,
- *  then the first element of collection is used as the initial value
- *  of <i>memo</i>.
- *
- *
- *     # Sum some numbers
- *     (5..10).reduce(:+)                             #=> 45
- *     # Same using a block and inject
- *     (5..10).inject { |sum, n| sum + n }            #=> 45
- *     # Multiply some numbers
- *     (5..10).reduce(1, :*)                          #=> 151200
- *     # Same using a block
- *     (5..10).inject(1) { |product, n| product * n } #=> 151200
- *     # find the longest word
- *     longest = %w{ cat sheep bear }.inject do |memo, word|
- *        memo.length > word.length ? memo : word
- *     end
- *     longest                                        #=> "sheep"
+ *     inject(symbol) -> obj
+ *     inject(initial_operand, symbol) -> obj
+ *     inject {|memo, element| ... } -> obj
+ *     inject(initial_operand) {|memo, element| ... } -> obj
+ *
+ *  Combines each successive element of +self+ with an accumulator.
```

Comment:
The explanation given here makes the method behavior sound like there are four separate, independent cases, and makes the documentation much longer than it needs to be.  The four cases aren't really independent, there are just 2 separate binary decisions based on the arguments:

1. Block or operator symbol
2. Initial operand or not

If an operator symbol is provided instead of a block, it is treated as the following block: `{|memo, element| memo.send(symbol, element)}`.

The memo argument to the block is the initial operand in the first iteration, and the result of the previous block iteration for subsequent iterations.

If in initial operand is not provided, the first element in the enumerable is treated as the initial operand, and the first block iteration is called with the second element in the enumerable as the element (second argument).

For empty enumerables, the initial operand is returned, or `nil` if there is no initial operand. For single element enumerables without an initial operand, the single element is returned. In all other cases, the return value is the return value of the last block iteration.

Is it possible refactor the documentation to make it shorter and still contain the same information?

Reply:
See comment below.